app/src/main/java/dev/shreyaspatil/foodium/data/repository/PostsRepository.kt

```diff
@@ -61,7 +59,7 @@ class PostsRepository @Inject constructor(
             override fun fetchFromLocal(): Flow<List<Post>> = postsDao.getAllPosts()
 
             override suspend fun fetchFromRemote(): Response<List<Post>> = foodiumService.getPosts()
-        }.asFlow().flowOn(Dispatchers.IO)
+        }.asFlow()
```

Comment:
@shubham08gupta Well, in this `flow` we are also fetching some data from the network. This `Dispatcher` we have used here for `flow` and not for Room database operation. Right?

Reply:
@PatilShreyas Just like the Room database provides in-built threading capability, the Retrofit(from v2.6.0) also provides in-built threading capability. You've already used a suspending Retrofit call which automatically makes it main-thread safe.

`suspend fun getPosts(): Response<List<Post>>`

There's no point in using `Dispatcher` in `flow` because now the background work is handled automatically.

Reply:
Yes, I agree with your point that ***Room*** and ***Retrofit*** task is handled which is also main-thread safe. But my concern is about `flow`. Flow is still being executed on a MAIN thread. Since it's *background* task I think `flow` should also be executed on ***IO*** thread.

---

Let me explain you. You can also test it.

Now remove `Dispatchers.IO` for `flow` and in `flow` body write a logging statement as-

```kotlin
Log.d("flowDispatcher", Thread.currentThread().toString())
```

Now run this app and see the output 🤔. You'll see it's output like `flowDispatcher: Thread[main,5,main]` which means `flow` is now executing on a MAIN thread which is useless because *this* is a repository and it doesn't require a MAIN thread to perform operations.

---

Let's repeat the same process for a IO. 🔍 

Now add `Dispatchers.IO` for `flow` and in `flow` keep logging statement as we did above and observe change-

You'll see output like `flowDispatcher: Thread[DefaultDispatcher-worker-1,5,main]` which will ensure you that flow 🌊  is not flowing on a MAIN thread.

**Reason:** Since this is a repository and it'll not include UI related tasks which don't require MAIN thread. As of now, Room and Retrofit operations are main-thread safe but suppose in future if we introduce any new operation here then it has to be main-thread safe. So I think it's not bad to use `Dispatchers.IO` here.

I hope you will get my point.

Reply:
Thanks for the clear explanation. 

> this is a repository and it doesn't require a MAIN thread to perform operations.

> Since this is a repository and it'll not include UI related tasks which don't require MAIN thread

I think there isn't any rule to execute Repository in the main thread. Please correct me if I'm wrong.

> but suppose in future if we introduce any new operation here then it has to be main-thread safe. So I think it's not bad to use Dispatchers.IO here.

We cannot be sure that we need `Dispatchers.IO` because we do not know what operation we are performing. Choosing a dispatcher will be based on the task that we want to perform(reading from a file/disk, sorting a list)
<img width="707" alt="Screen Shot 2020-09-03 at 18 21 19" src="https://user-images.githubusercontent.com/7660565/92096794-456fd380-ee12-11ea-807f-20bcec5e1389.png">
Ref - [https://medium.com/androiddevelopers/coroutines-on-android-part-i-getting-the-background-3e0e54d20bb](https://medium.com/androiddevelopers/coroutines-on-android-part-i-getting-the-background-3e0e54d20bb)

I think executing the `flow` just on a different thread(for future safety) is an overhead of context switching which is not required. 
If you still want to use a Dispatcher, you can. 🙏 🙌

Reply:
> I think there isn't any rule to execute Repository in the main thread. Please correct me if I'm wrong.

Yes, you're right.

> We cannot be sure that we need `Dispatchers.IO` because we do not know what operation we are performing. Choosing a dispatcher will be based on the task that we want to perform(reading from a file/disk, sorting a list) 
> I think executing the `flow` just on a different thread(for future safety) is an overhead of context switching which is not required.

Hmm.. this is also right. We can choose Dispatcher whenever it's required in future. I completely agree with you. Thanks for the explanation 😃. I'm happy to merge this.